src/Cms/Cors.php

```diff
@@ -0,0 +1,236 @@
+			return [];
+		}
+
+		$headers = [];
+
+		// determine allowed origin
+		$allowOrigin = $this->allowOrigin();
+
+		// no origin match found
+		if ($allowOrigin === null) {
+			return [];
+		}
+
+		$headers['Access-Control-Allow-Origin'] = $allowOrigin;
+
+		$this->addVaryHeader($headers, $allowOrigin);
+		$this->addCredentialsHeader($headers, $allowOrigin);
+		$this->addExposeHeaders($headers);
+
+		// add preflight-specific headers
+		if ($preflight === true) {
+			$this->addPreflightHeaders($headers);
+		}
+
+		return $headers;
+	}
+
+	/**
+	 * Determines the allowed origin based on config and request
+	 */
+	protected function allowOrigin(): string|null
+	{
+		$requestOrigin = $this->kirby->request()->header('Origin');
+		$configOrigin  = $this->config['allowOrigin'] ?? '*';
+
+		return match (true) {
+			$configOrigin === '*' => '*',
+			$requestOrigin === null => null,
+			is_string($configOrigin) => strcasecmp($configOrigin, $requestOrigin) === 0 ? $requestOrigin : null,
+			default => $this->matchOriginFromArray($configOrigin, $requestOrigin)
+		};
+	}
+
+	/**
+	 * Matches the request origin against an array of allowed origins
+	 *
+	 * @param array $configOrigin Array of allowed origins
+	 * @param string $requestOrigin Origin from the request header
+	 */
+	protected function matchOriginFromArray(array $configOrigin, string $requestOrigin): string|null
+	{
+		foreach ($configOrigin as $origin) {
+			if (strcasecmp($origin, $requestOrigin) === 0) {
+				return $requestOrigin;
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Adds the Vary header for cache control
+	 *
+	 * @param array $headers Headers array (passed by reference)
+	 * @param string $allowOrigin Allowed origin value
+	 */
+	protected function addVaryHeader(array &$headers, string $allowOrigin): void
+	{
+		// response varies by origin for non-wildcard origins
+		if ($allowOrigin !== '*') {
+			$currentVary = $this->kirby->request()->header('Vary');
+			$headers['Vary'] = $currentVary ?? 'Origin';
+		}
+	}
+
+	/**
+	 * Adds the credentials header if configured
+	 *
+	 * @param array $headers Headers array (passed by reference)
+	 * @param string $allowOrigin Allowed origin value
+	 */
+	protected function addCredentialsHeader(array &$headers, string $allowOrigin): void
+	{
+		$allowCredentials = $this->config['allowCredentials'] ?? false;
+
+		if ($allowCredentials === true) {
+			// wildcard origins cannot be used with credentials
+			if ($allowOrigin === '*') {
+				throw new InvalidArgumentException(
+					message: 'Cannot allow credentials when using wildcard origin (*)'
+				);
+			}
+
+			$headers['Access-Control-Allow-Credentials'] = 'true';
+		}
+	}
+
+	/**
+	 * Adds headers to expose custom headers to the client
+	 *
+	 * @param array $headers Headers array (passed by reference)
+	 */
+	protected function addExposeHeaders(array &$headers): void
+	{
+		$exposeHeaders = $this->config['exposeHeaders'] ?? [];
+
+		if (empty($exposeHeaders) === false) {
```

Comment:
`Todo`: We are trying to get away from the `empty()` function (see [this post](https://dev.to/klnjmm/never-use-empty-function-in-php-4pb0)).

As all four cases in this file check whether the input is either a non-empty string or a non-empty array, we could add a `protected function configIsEmpty(mixed $input)` that does:

```php
return (
	(is_array($input) === true && $input !== []) ||
	(is_string($input) === true && $input !== '')
);
```

This would harden the logic against invalid configuration.